sw/cheri/common/platform-pinmux.hh.tpl

```diff
@@ -1,195 +1,201 @@
   }
+}
 
-  /**
-   * For a given output pin, selects a given block output to use for that pin
-   * via the pin multiplexer.
-   *
-   * @param output_pin The output pin to pinmux.
-   * @param option The option to select for that pin. This value should be
-   * less than the value returned by `output_pin_options` for the given pin.
-   *
-   * The meanings of these selections can be found in the documentation:
-   * https://lowrisc.github.io/sonata-system/doc/ip/pinmux.html
-   */
-  bool output_pin_select(OutputPin output_pin, uint8_t option) {
-    if (option >= output_pin_options(output_pin)) {
-      Debug::log("Selected option is not valid for this pin.");
-      return false;
-    }
-    uint16_t registerOffset   = static_cast<uint16_t>(output_pin);
-    registers[registerOffset] = (1 << option);
-    return true;
-  }
+/**
+ * A handle to a sink configuration register. This can be used to select
+ * the source of the handle's associated sink.
+ */
+template <typename SinkEnum>
+struct Sink {
+  CHERI::Capability<volatile uint8_t> reg;
+  const SinkEnum sink;
 
   /**
-   * For a given block input, selects a pin to use for that input via the pin
-   * multiplexer.
+   * Select a source to connect to the sink.
    *
-   * @param block_input The block input to pinmux.
-   * @param option The option to select for that block input. This value
-   * should be less than the value returned by `block_input_options` for the
-   * given block input.
+   * To see the sources available for a given sink see the Sonata system
+   * documentation:
+   * https://lowrisc.github.io/sonata-system/doc/ip/pinmux/
    *
-   * The meanings of these selections can be found in the documentation:
-   * https://lowrisc.github.io/sonata-system/doc/ip/pinmux.html
+   * Note, source 0 disconnects the sink from any source disabling it,
+   * and source 1 is the default source for any given sink.
    */
-  bool block_input_select(BlockInput block_input, uint8_t option) {
-    if (option >= block_input_options(block_input)) {
-      Debug::log("Selected option is not valid for this block.");
+  bool select(uint8_t source) {
+    if (source >= sources_number(sink)) {
+      Debug::log("Selected source not within the range of valid sources.");
       return false;
     }
-    uint16_t registerOffset   = static_cast<uint16_t>(block_input);
-    registers[registerOffset] = (1 << option);
+    *reg = 1 << source;
     return true;
   }
 
-  /**
-   * A constructor for the SonataPinmux driver, which takes a bounded
-   * capability to the pinmux registers. This should be replaced with
-   * an appropriate `MMIO_CAPABILITY` call in the version of the driver
-   * that runs in CHERIoT RTOS, rather than baremetal.
-   */
-  SonataPinmux(volatile uint8_t *registers) : registers(registers) {}
+  /// Disconnect the sink from all available sources.
+  void disable() { *reg = 0b01; }
+
+  /// Reset the sink to it's default source.
+  void default_selection() { *reg = 0b10; }
```

Comment:
Nit: I think I prefer using `(1 << 0)` and `(1 << 1)` in place of the literals `0b01` and `0b10` just so it's clearer to anyone reading how the `disable` and `default_selection` methods are related to the `select` method above, but that's up to your preference.